File: ts/wasm/__tests__/detailed/hive_assertion.ts
import { expect } from '@playwright/test';

import { test } from '../assets/jest-helper';
import type { WaxProtocolAssertionError } from '../../dist/bundle';

import type { claim_account, operation } from '../../dist/bundle';
import type { IWasmGlobals, IWaxGlobals } from '../assets/globals';

test.describe('Wax tests verifying unique assertion exceptions from hive', () => {

  const txValidate = async ({ chain, wax }: { chain: IWaxGlobals["chain"]; wax: IWaxGlobals["wax"] }, testedOp: operation) => {
    // Create transaction
    const tx = await chain.createTransaction();
    tx.pushOperation(testedOp);

    try {
      tx.validate();
    }
    catch (e) {
      if(e && typeof e === "object") {
        const error: object = e as object;
        if(e instanceof wax.WaxProtocolAssertionError) {
          const caughtAssertion: WaxProtocolAssertionError = error as WaxProtocolAssertionError;
          // To extract assertion expression we need object form of message json.
          const objectMsg = JSON.parse(caughtAssertion.message);
          return {
            detectedError: {
              source: "protocol",
              expression: objectMsg.extension.assertion_expression || "Unknown assertion expression",
              hash: caughtAssertion.assertionHash
            }
          };
        } else {
          const errorStr = JSON.stringify(error);
          return { detectedError: {message: errorStr} };
        }
      }

      throw new Error("Unexpected error type caught: " + e);
    }

    throw new Error("No error detected");
  };

  test('Expecting assertion evaluating invalid operation', async ({ waxTest }) => {
    // Validate invalid claim_account operation
    const testedOp: claim_account = {
      creator: "user123",
      fee: { nai: "@@000000013", amount: "1", precision: 3 },
      extensions: []
    };
    const op: operation = { claim_account_operation: testedOp };
    const retVal = await waxTest(txValidate, op);
    expect(retVal.detectedError).toStrictEqual({
      source: "protocol",
      expression: "is_asset_type( fee, HIVE_SYMBOL ) && \"Account claiming fee must be HIVE\"",
      hash: "14687464191050907756"
    });

    // Validate another invalid operation to trigger a different assertion ...
  });
});

test.describe('WASM Protocol assertions', () => {
  const validateOperation = async ({ protocol, provider }: { protocol: IWasmGlobals["protocol"]; provider: IWasmGlobals["provider"] }, testedOp: operation) => {
    // Create transaction
    const handle = protocol.cpp_create_operation_handle(testedOp, false);

    try {
      protocol.cpp_op_validate(handle);
    }
    catch (e) {
      console.log(`WaxBaseApi: C++ exception thrown during initialization: ${e}`);
      const d = provider.getExceptionMessage(e);
      console.log(`Received error details from getExceptionMessage:\nexception-type: ${d[0]},\nexception-message: ${d[1]}`);
      if(e && typeof e === "object") {
        const error: object = e as object;
        try {
          // To extract assertion expression we need object form of message json.
          const objectMsg = JSON.parse(d[1]);
          return {
            detectedError: {
              type: d[0],
              expression: objectMsg.extension.assertion_expression || "Unknown assertion expression",
              hash: objectMsg.assert_hash || "Unknown assertion hash"
            }
          };
        }
        catch (e2) {
          const errorStr = JSON.stringify(error);
          return { detectedError: {message: errorStr} };
        }
      }

      throw new Error("Unexpected error type caught: " + e);
    }

    throw new Error("No error detected");
  };

  test('Testing getExceptionMessage as wasmTest', async ({ wasmTest }) => {
    // Validate invalid claim_account operation
    const retVal = await wasmTest(validateOperation, {
      "type": "claim_account_operation",
      "value": {
        "creator": "user123",
        "fee": { "nai": "@@000000013", "amount": "1", "precision": 3 }
      }
    } as operation);

    expect(retVal.detectedError).toStrictEqual({
      type: "cpp::wax_protocol_assertion",
      expression: "is_asset_type( fee, HIVE_SYMBOL ) && \"Account claiming fee must be HIVE\"",
      hash: "14687464191050907756"
    });
  });
});

File: rust/wax/tests/detailed/hive_assertion.rs
// Rust port of `ts/wasm/__tests__/detailed/hive_assertion.ts`.
//
// Tests appear in TS source order. Each Rust test has a `// TS line N` comment
// pointing back to the TS original.
//
// TS NOTE: TS wraps hive assert exceptions in a typed `WaxAssertionError`
// (category / raw / assertHash); the Rust `WaxError` carries the raw fc
// exception JSON as its message, so the tests parse that JSON and assert the
// same fields.
//
// TS NOTE: the TS fixture pins the pre-refactor assertion
// `is_asset_type( fee, HIVE_SYMBOL ) && "Account claiming fee must be HIVE"`
// (hash 14687464191050907756). The current hive submodule routes the check
// through `validate_asset_type` (`HIVE_PROTOCOL_ASSET_ASSERT_WITH_SUBJECT`),
// so the expression is `is_asset_type( asset, symbol)` (hash
// 7633970631494007356)
// with the human-readable reason carried in `data.context` instead.

use serde_json::Value;

use wax::proto::{self, operation::Value as OperationValue};

use crate::common::wax_test;

/// The invalid operation both TS tests validate: an account-claiming fee
/// expressed in HBD (must be HIVE).
fn invalid_claim_account() -> OperationValue {
    OperationValue::ClaimAccountOperation(proto::ClaimAccount {
        creator: "user123".into(),
        fee: proto::Asset {
            nai: "@@000000013".into(),
            amount: "1".into(),
            precision: 3,
        },
        extensions: Vec::new(),
    })
}

/// Parses the fc exception JSON carried by the validation error and returns
/// `(category, assertion_expression, assert_hash, context)`.
fn assertion_details(error: &str) -> (String, String, String, String) {
    let parsed: Value =
        serde_json::from_str(error).expect("assertion error is JSON");

    assert_eq!(parsed["name"], "assert_exception");

    (
        parsed["stack"][0]["data"]["category"]
            .as_str()
            .unwrap_or_default()
            .to_string(),
        parsed["extension"]["assertion_expression"]
            .as_str()
            .expect("assertion_expression")
            .to_string(),
        parsed["assert_hash"]
            .as_str()
            .expect("assert_hash")
            .to_string(),
        parsed["stack"][0]["data"]["context"]
            .as_str()
            .unwrap_or_default()
            .to_string(),
    )
}

// TS line 44: "Expecting assertion evaluating invalid operation" — the
// transaction-level validation path.
#[test]
fn assertion_evaluating_invalid_operation() {
    wax_test(None, |ctx| {
        let mut tx = ctx
            .base
            .create_transaction_with_tapos(
                "04c507a8c7fe5be96be64ce7c86855e1806cbde3",
                "2023-11-09T21:51:27",
            )
            .expect("create_transaction_with_tapos");
        tx.push_operation(ctx.base.create_operation(invalid_claim_account()));

        let error = tx.validate().expect_err("expected validation error");
        let (category, expression, hash, context) =
            assertion_details(&error.to_string());

        assert_eq!(category, "protocol");
        assert_eq!(expression, "is_asset_type( asset, symbol)");
        assert_eq!(hash, "7633970631494007356");
        assert_eq!(context, "Account claiming fee must be HIVE");
    });
}

// TS line 100: "Testing getExceptionMessage as wasmTest" — the raw
// operation-level protocol validation path.
#[test]
fn get_exception_message_from_operation_validation() {
    wax_test(None, |ctx| {
        let operation = ctx.base.create_operation(invalid_claim_account());

        let error =
            operation.validate().expect_err("expected validation error");
        let (category, expression, hash, context) =
            assertion_details(&error.to_string());

        assert_eq!(category, "protocol");
        assert_eq!(expression, "is_asset_type( asset, symbol)");
        assert_eq!(hash, "7633970631494007356");
        assert_eq!(context, "Account claiming fee must be HIVE");
    });
}
